Back-end and UI for off-chain offers

franckc opened the pull request:
Refactored the logic used by the makeOffer processor to check whether the flag "enabledOffchainOffers" is set on a shop's config to determine if an offer should be recorded on or off-chain. If the offer is to be recorded off-chain, then A row in the orders table is immediately inserted and notifications (email, discord) are sent.

The orders table DB schema got updated:
 - Changed the type of the currently unused "status" column from string to enum to reflects the status of the order 
 - Renamed the "statusStr" column to "offerStatus"
 - Added a new column offer_id, populated for on-chain offers

Notes:
 1. Initially we had discussed using the IPFS hash of the offer as unique order id. But that's pretty long (46 chars) and it would not make sense to use it in emails and whatever communication between the buyer and seller. So I'm proposing we use the following human readable format for order_id: <networkId>-<shopId>-<8 characters random Id>. Please let me know feedback on this. networkId and shopId are probably not necessarily mandatory so we could drop them. My thinking was it might be handy to have them in there to clear any confusion between staging/prod and also between orders from different shops. Another thing to think about is whether or not it's ok to expose the shopId as a plain Id. It reveals the number of shops in the system. We could encrypt it. But I'm thinking since we are already using plain shopIds in other places (like in the API data exchanged between the UI and the server), it's probably ok.
 1. Backend/utils currently has a mix of utility methods but also business logic. I think it would be cleaner to separate out business logic in a dedicated directory. So I added a new directory backend/logic and put in there the new logic code for orders. Moving forward I think it would be nice to put all the business logic in there.

Still TODO in this PR:
 - Determine if we need to backfill legacy rows in the orders table that were using the full-qualified offerId as orderId and have them use the new order id format. This may cause confusion for users as they received emails with the legacy format.
 - Check to make sure that setting the enabledOffchainOffers flag in a shop won't cause issue for legacy orders on that shop, especially if any order has not been fulfilled yet.
 - Integrate with the UI, in particular to properly handle the change of order.statusStr -> order.offerStatus

TODO in a separate PR:
 - Expose a new endpoints for canceling an off-chain order.

franckc commented:
@nick @shahthepro @mikeshultz This PR is not completely finished (see list of TODOs in comments), but I thought I would send it out to get early feedback on the approach before I finalize it. Lmk any feedback. Thanks!

mikeshultz commented:
Can we/I get a high level overview of what the off-chain payment process looks like?  I'm not really sure what the intention of this PR is other than "off-chain payments."

> Initially we had discussed using the IPFS hash of the offer as unique order id. But that's pretty long (46 chars) and it would not make sense to use it in emails and whatever communication between the buyer and seller. So I'm proposing we use the following human readable format for order_id: --<8 characters random Id

Could use a truncated IPFS hash, then.  Last N characters, for example.  Just throwing it out there.

> Backend/utils currently has a mix of utility methods but also business logic. I think it would be cleaner to separate out business logic in a dedicated directory. So I added a new directory backend/logic and put in there the new logic code for orders. Moving forward I think it would be nice to put all the business logic in there.

Then we end up with the AC design where everything ends up under `/logic` instead of `/utils`.  Instead, maybe we need new modules with discrete purposes.  For example, `/offers` for any offer-related logic. I'm personally a fan of more directories to break up the files.

franckc commented:
@mikeshultz Here is the flow involved in an off-chain payment:
 1. Buyer makes a payment with one of the supported payment type (stripe, paypal, direct)
 2. Store front-end calls back-end API (ex. in case of a direct crypto payment) or external payment service (paypal, stripe) calls backend webhook to process the payment
 3. Back-end enqueues a job in the "makeOffer" queue
 4. Job gets processed.  If the shop's config flag "offchainOffersEnabled" is set, then the order gets created in the DB but is not recorded on the blockchain. If the flag is not set, the offer is recorded on the blockchain by sending a "createOffer" transaction to the marketplace contract. When that transaction gets mined, the listener will pick up the "OfferCreated" event and process the order.

Regarding using the first N characters if the IPFS hash of the offer, it's definitely a possibility. I was not sure about the risk of collision. But my main concern is more related to the non human readable aspect of IPFS hashes, especially the combination of lower/upper case and the possibility of very hard to distinguish characters (ex: O00 or Il) next to each other...

mikeshultz commented:
Thanks for the overview, will try and review today.

> But my main concern is more related to the non human readable aspect of IPFS hashes, especially the combination of lower/upper case and the possibility of very hard to distinguish characters (ex: O00 or Il) next to each other...

We doing phone support now?

franckc commented on backend/queues/makeOfferProcessor.js:
```diff
@@ -42,9 +43,213 @@ function attachToQueue() {
+  //    could be getting processed.
+  //  - The processing a job could get interrupted while waiting for the tx to get mined.
+  //    For example due to a server maintenance or a crash.
+  // TODO: In order to avoid any possible race condition, hold a lock on a DB row
```
yes good point. updated the comment.

franckc commented on backend/logic/order.js:
```diff
@@ -0,0 +1,160 @@
+  }
+  if (data.referrer) {
+    orderObj.referrer = util.toChecksumAddress(data.referrer)
+    orderObj.commissionPending = Math.floor(data.subTotal / 200)
```
Yes, all this code was just legacy that I moved to this new file.
I agree we need to tighten up the offer validation. Let's handle this as a separate PR if that's ok. It's already being tracked as issue #17

franckc commented on backend/logic/order.js:
```diff
@@ -0,0 +1,160 @@
+
+  // Extract the optional paymentCode data from the offer.
+  // It is populated for example in case of a Credit Card payment.
+  const paymentCode = offer.paymentCode || null
```
It's a unique ID assigned by the back-end when a payment is initiated. It allows to track the various steps involved in the order processing lifecycle.

franckc commented on backend/logic/order.js:
```diff
@@ -0,0 +1,160 @@
+    capitalization: 'uppercase',
+    length: 8
+  })
+  return `${shop.networkId}-${shop.id}-${randomId}`
```
If we decide to use listingId, I'd vote for always using the server listingId. So that in the future we don't require a listingId when a merchant sets up a new shop (that's a high friction point).

A downside is that it will add an extra step (e.g. sending a transaction to create a listing + wait for it to get mined) when a new deployer system gets setup (ex: via GCP marketplace). But that's probably ok given one could expect that a deployer setup is a fairly technical process.

franckc replied:
We decided to use <netId>-<contractVersion>-<listingId>-<shopId>-<randomId> for the fully qualified form.
For the short form used as a reference number for external communication, we'll just use <randomId>.

franckc commented:
@mikeshultz Sorry I missed your earler comment about /logic. Fair point that a lot of stuff could end up in there. How about having subdirectories under /logic? like
/logic/offer
/logic/stripe
/logic/paypal
/logic/discount
etc...

franckc commented:
Updated this PR to include changes to the UI.

Also made a change to the DB schema: renamed the order.status column to order.payment_status for clarity.

Still TODO:
 [ ] decide on whether or not we should use the short version of the orderId for all pages/email facing user and merchant. Right now it is using the full qualified order Id.
 [ ] determine how to populate network.listingId. Best would be via super-admin UI at setup but I might defer this to a separate PR since this one is getting huge.
 [ ] decide if we should populate shops.listingId with the network's listingId at shop creation or just leave it empty.
 [ ] write a script to populate all orders.payment_status to Paid/Refunded based on offer_status
 [ ] review impact of this change for legacy orders already in the system to make sure they don't break.
